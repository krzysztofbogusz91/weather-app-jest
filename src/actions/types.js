export const FETCH_WEATHER = 'FETCH_WEATHER';
export const FETCH_FORECAST = 'FETCH_FORECAST';
export const USER_CORDS = 'USER_CORDS';
export const CLEAR = 'CLEAR'
export const STATUS_PENDING = 'STATUS_PENDING';
export const STATUS_SUCCESS = 'STATUS_SUCCESS';
export const STATUS_ERROR = 'STATUS_ERROR'

