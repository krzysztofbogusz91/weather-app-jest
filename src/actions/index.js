import * as types from './types';
import { term } from '../helpers/check_search';

const API_KEY = '54df0301d1505a0aee49fe3b417ecd92';
const urlMain = `https://api.openweathermap.org/data/2.5/`;

export const clearFetch = () => {
    return {
        type: types.CLEAR,
    }
}

export const pendingFetch = () => {
    return {
        type: types.FETCH_PENDING,
        isLoading: true
    }
}

export const updateInput = term =>{
    return {
        type: types.INPUT_VAL,
        payload: term
    }
}

export const fetchWeather = data => dispatch => {

    let search = term(data);
    const url = `${urlMain}weather?${search}&appid=${API_KEY}`;



    return fetch(url)
        .then(res => res.json())
        .then(data => {
            dispatch({
                type: types.FETCH_WEATHER,
                payload: data,
                error: false,
                isLoading: false
            })
        }
        ).catch(err => {
            dispatch({
                type: types.FETCH_ERR,
                error: true,
                isLoading: false
            })
        });
}

export const fetchForecast = data => dispatch => {

    let search = term(data);
    const url = `${urlMain}forecast?${search}&appid=${API_KEY}`;

    return fetch(url)
        .then(res => res.json())
        .then(data => {
            dispatch({
                type: types.FETCH_FORECAST,
                payload: data,
                error: false,
                isLoading: false
            })
        }
        ).catch(err => {
            dispatch({
                type: types.FETCH_ERR,
                error: true,
                isLoading: false
            })

        });
}

export const fetchUser = () => dispatch => {

    // if (navigator.geolocation) {
    //     /* geolocation is available */
    //     console.log('geo yes')
    //     navigator.geolocation.getCurrentPosition((pos)=>{
    //         const crd = pos.coords
    //         dispatch({type: types.USER_CORDS, payload: {lat: crd.latitude, lng: crd.longitude} })
    //     });
    //   } else {
    /* geolocation IS NOT available */

    const key = 'AIzaSyB0jxnnSPREDEgoCvskYDxjJPJ2nSAzDaA'
    const url = `https://www.googleapis.com/geolocation/v1/geolocate?key=${key}`;

    return fetch(url, { method: 'POST' })
        .then(res => res.json())
        .then(data => {
            dispatch({
                type: types.USER_CORDS,
                payload: data.location
            })
        }
        ).catch(err => err);
    //   } end if else
}

export const autoComplete = term => dispatch => {
    
    console.log('term: ​', term );
    const data = [
    { city: "Katex", state: "Saki", country: "Azerbaijan" },

    { city: "Texcalyacac", state: "Mexico", country: "Mexico" },

    { city: "Texcoco", state: "Mexico", country: "Mexico" },

    { city: "Texmelucan", state: "Puebla", country: "Mexico" },

    { city: "Atexcatzingo", state: "Tlaxcala", country: "Mexico" },
]

   dispatch({
        type: types.AUTO_COM,
        payload: data
    })
    console.log('dispached')



    // console.log('wywo')

    // const key = `Yh5PIaZSUSmshH4egy13RmkXdUEHp1czYGmjsnWRfgrt90PVus`
    // const url = `https://andruxnet-world-cities-v1.p.mashape.com/?query=${term}&searchby=city`;

    // const myHeaders = new Headers({
    //     "Accept": "application/json",
    //     "X-Mashape-Key": key
    // });

    // const myInit = {
    //     method: 'GET',
    //     headers: myHeaders,
    //     mode: 'cors',
    //     cache: 'default'
    // };

    // const myRequest = new Request(url, myInit);

    // console.log('build​', );

    // return fetch(myRequest)
    //     .then(res => res.json())
    //     .then(data => {
    //         console.log(data)
    //         dispatch({
    //             type: types,
    //             payload: data
    //         })
    //     }
    //     ).catch(err => err);
}

// const data = [
//     { city: "Katex", state: "Saki", country: "Azerbaijan" },

//     { city: "Texcalyacac", state: "Mexico", country: "Mexico" },

//     { city: "Texcoco", state: "Mexico", country: "Mexico" },

//     { city: "Texmelucan", state: "Puebla", country: "Mexico" },

//     { city: "Atexcatzingo", state: "Tlaxcala", country: "Mexico" },
// ]