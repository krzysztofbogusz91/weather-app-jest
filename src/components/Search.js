import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchWeather, fetchForecast, clearFetch, pendingFetch, autoComplete, updateInput } from '../actions/index';
import { Table } from '../containers/Table';
import { ErrorComponent } from '../containers/ErrorComponent'
import MapContainer from "./MapContainer";
import { Loader } from '../containers/Loader';
import AutoCom from './AutoCom';


export class Search extends Component {
    constructor(props) {
        super(props);
        this.state = {
            renderTable: false,
        }
    }

    inputChange = (e) => {
        //NEED TO ADD TESTS TO THIS
        if (this.props.inputVal.length === 3) {
            this.props.autoComplete(this.props.inputVal);
        }

        this.props.updateInput(e.target.value);

        //this.props.inputVal.length === 0 ? this.setState({ renderTable: false }) : null
    }

    buttonOnClick = (e) => {

        //ADD SUBMIT VALIDATION AND TESTS
        e.preventDefault()

        //Show loader when data is pending
        this.props.pendingFetch()

        const cord = this.props.inputVal;

        this.props.fetchForecast(cord);
        this.props.fetchWeather(cord);

        this.setState({
            renderTable: true,
        })
    }

    componentWillUnmount() {
        //prevents issue of data disply afeter route change
        this.props.clearFetch();
    }

    render() {
        const renderTable = this.state.renderTable && !this.props.isLoading;
        const coord = this.props.today.length > 0 && this.props.today[0].coord
        const lat = coord ? this.props.today[0].coord.lat : 0;
        const lon = coord ? this.props.today[0].coord.lon : 0;
        //console.log(this.props)
        //console.log(this.state.renderTable)
        //console.log('â€‹Search -> render -> ', this.props.auto );
        //console.log(this.props.isLoading)

        //TODO MAP RENDERS ON FIRST QUERY ISSUE
        //TODO => WHEN TYPING IN INPUT SHOW TOOLTIP ASKING TO PASS COUNTRY CODE AFTER COMA, or even on clik pass chosen one
        return (
            <div className="mt-5">
                <form action="submit">

                    <input
                        value={this.props.inputVal}
                        onChange={this.inputChange}
                        type="text"
                        className='search-fetch form-control' />
                    {this.props.auto.length > 0 ? <AutoCom /> : null}
                    <button type="submit" className="mt-2 btn btn-info form-control search" onClick={this.buttonOnClick}>
                        Search
                        </button>
                </ form>

                <div className='current-weather mt-3'>
                    {this.props.isLoading ? <Loader /> : null}
                    {renderTable && this.props.error ? <ErrorComponent /> : null}
                    {renderTable && !this.props.error ?
                        <div>
                            <MapContainer lat={lat} lng={lon} />
                            <Table
                                type="today"
                                table={this.props.today} />
                            <Table
                                type="forecast"
                                table={this.props.weather} />
                        </div>
                        : null}
                </div>
            </div>
        );
    }
}

Search.propTypes = {
    fetchWeather: PropTypes.func.isRequired,
    fetchForecast: PropTypes.func.isRequired,
    today: PropTypes.any,
    weather: PropTypes.any,
    cords: PropTypes.any,
};

Search.defaultProps = {
    today: [],
    weather: [],
    cords: {},
}

export default connect(state => ({ ...state }), { fetchWeather, fetchForecast, clearFetch, pendingFetch, autoComplete, updateInput })(Search);
