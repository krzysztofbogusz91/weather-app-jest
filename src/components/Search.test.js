import React from 'react';
import { Search }  from './Search'
import set from '../enzyme_set_up';
import { shallow, mount } from 'enzyme';
import { fetchWeather, fetchUser, fetchForecast } from '../actions/index';


const mockfetchWeather = jest.fn()
const mockfetchUser = jest.fn(() => { props.cords = { cords: "cor" } })
const mockfetchForecast = jest.fn();
const props = { today: [], weather: [], cords: {}, fetchUser: mockfetchUser, fetchForecast: mockfetchForecast, fetchWeather: mockfetchWeather }

describe('Search', () => {

    //USE MOUNT TO TEST PROPS VALUES AND CHANGES
    const search = shallow(<Search {...props} />)

    it('renders without crashing', () => {
        expect(search).toMatchSnapshot();
    });

    it('should have initial empty state ', function () {
        const bar = search.find('.search-fetch');
        expect(search.state().search).toEqual("");
    });

    it('should have input search bar that sets state ', function () {
        const bar = search.find('.search-fetch');
        bar.simulate('change', { target: { value: "Lond" } })
        expect(search.state().search).toEqual("Lond");
    });

    it('should have search button', function () {
        const button = search.find('.search');
        
        expect(button.exists()).toBe(true);
    });

  
    //NOT NECESSARY => just other way than above
    it('should call inputChage method and c', () => {
        search.instance().inputChange({ target: { value: "Lod" } })
        expect(search.state().search).toEqual("Lod");
    });

    it('fetchUser should update props', () => {
        expect(search.props().cords).not.toEqual({})
    });

    it('should not show table when list length > 0', () => {
        expect(search.find('.search-weather').children().length).toEqual(0)
    });
})

describe('Mount Search', () => {

    const search = mount(<Search {...props} />)

    it('On button click should call feches with search state', function () {
        search.setState({
            search: "Lond"
        })
        const button = search.find('.search').simulate('click');
        expect(search.props().fetchForecast).toHaveBeenCalledWith("Lond");
        expect(search.props().fetchWeather).toHaveBeenCalledWith("Lond");
    });
});

