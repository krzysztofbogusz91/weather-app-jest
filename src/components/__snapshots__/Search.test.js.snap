// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    cords={Object {}}
    fetchForecast={[Function]}
    fetchUser={[Function]}
    fetchWeather={[Function]}
    pendingFetch={[Function]}
    today={Array []}
    weather={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          action="submit"
>
          <input
                    className="search-fetch form-control"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <button
                    className="mt-2 btn btn-info form-control search"
                    onClick={[Function]}
                    type="submit"
          >
                    Search
          </button>
</form>,
        <div
          className="current-weather mt-3"
>
          
</div>,
      ],
      "className": "mt-5",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "action": "submit",
          "children": Array [
            <input
              className="search-fetch form-control"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <button
              className="mt-2 btn btn-info form-control search"
              onClick={[Function]}
              type="submit"
>
              Search
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "search-fetch form-control",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
              "className": "mt-2 btn btn-info form-control search",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            null,
            null,
            null,
          ],
          "className": "current-weather mt-3",
        },
        "ref": null,
        "rendered": Array [
          null,
          null,
          null,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            action="submit"
>
            <input
                        className="search-fetch form-control"
                        onChange={[Function]}
                        type="text"
                        value=""
            />
            <button
                        className="mt-2 btn btn-info form-control search"
                        onClick={[Function]}
                        type="submit"
            >
                        Search
            </button>
</form>,
          <div
            className="current-weather mt-3"
>
            
</div>,
        ],
        "className": "mt-5",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "action": "submit",
            "children": Array [
              <input
                className="search-fetch form-control"
                onChange={[Function]}
                type="text"
                value=""
/>,
              <button
                className="mt-2 btn btn-info form-control search"
                onClick={[Function]}
                type="submit"
>
                Search
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "search-fetch form-control",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
                "className": "mt-2 btn btn-info form-control search",
                "onClick": [Function],
                "type": "submit",
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              null,
              null,
              null,
            ],
            "className": "current-weather mt-3",
          },
          "ref": null,
          "rendered": Array [
            null,
            null,
            null,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
